### PR TITLE
OPUSVIER-3657 Refactoring of logging setup

### PR DESCRIPTION
**Pull request by kaustabhbarman**
I have created the `getInstance()` method that will open and use the specified log file. I have tried to create two methods `setLog()` and `getLog()`.  `setLog()` needs a _**logName**_ and an optional _**format**_ . It will use any specified configuration parameter format in _application.ini_ like _**logging.log.translation.file.format = ' '**_ and _**logging.log.level = ''**_. For `getLog()`, I am not sure what exactly to look for here. Should I look for a log file? For now, it will return the logger from `setLog()`.

### CODE REVIEW

File: src/Log/LogService.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+	public static $logfile;
+	public static $logger;
+
+	public static function getInstance($logFileName){
+
+	    $registry = \Zend_Registry::getInstance();
+	    $config = $registry->get('Zend_Config');
+
+        $logfilePath = $config->workspacePath . '/log/' . $logFileName;
```
Comment (j3nsch) [inline src/Log/LogService.php:48]: Getting the path for the log files should be encapsulated in a separate function. The should also be a function for setting the path. That will be useful for testing. In the unit test we can set a temporary path and check if loggers are created as expected.

File: test/LogTest.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ * terms of the GNU General Public License as published by the Free Software
+ * Foundation; either version 2 of the Licence, or any later version.
+ * OPUS is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details. You should have received a copy of the GNU General Public License
+ * along with OPUS; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ * @category    Test
```
Comment (j3nsch) [inline test/LogTest.php:27]: Please update the header.

File: test/LogTest.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        $this->logger = $this->logInstance->setLog('INFO');
+        $this->doesNotPerformAssertions();
+    }
+
+    public function testGetLog(){       //not sure what exactly to assert
+        $log = $this->logInstance->getLog();
+        $this->assertSame($log, $this->logger);
+    }
+
+}
```
Comment (j3nsch) [inline test/LogTest.php:61]: I don't think these files pass the composer cs-check right now. I am not sure why the Github Actions aren't running for your pull request. Maybe a security feature by GitHub because it is your first PR to this project?

File: test/LogTest.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ * @copyright   Copyright (c) 2018, OPUS 4 development team
+ * @license     http://www.gnu.org/licenses/gpl.html General Public License
+ */
+
+namespace OpusTest;
+
+
+use Opus\Log\LogService;
+
+class LogTest extends \PHPUnit_Framework_TestCase
```
Comment (j3nsch) [inline test/LogTest.php:39]: I think we need at least the following tests. 

* creating default log
* creating default log with different name
* creating custom log
* getting default log
* getting custom log
* trying to get unknown log
* trying all the options
* ...

File: src/Log/LogService.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        \Zend_Registry::set('Zend_Log', $logger);
+
+        $logger->debug('Logging initialized');
+        self::$logger = $logger;
+        return $logger;
+    }
+
+	public function getLog($logLevelName){      //not sure what exactly to return here
```
[the remark this answers is not in the document]
Comment (j3nsch) [reply src/Log/LogService.php:124]: We do not distinguish between logs by Log level. We have multiple logs that may have different logs levels but we use a name to get the logger we want, e.g. 'default', 'doi', 'translation', ...

The function should check if a logger with that name already exists. If not it should try to create a logger from the configuration. If that fails it should throw an exception, UnknownLogger. Modules that want to use a new custom 
logger would add the createLog call to their Bootstrap process. At least that is what I am thinking right now, until we run into problems.

[the remark this answers is not in the document]
Comment (j3nsch) [reply src/Log/LogService.php:124]: When I am saying creating a logger from "configuration" it is something that could be added in the next step. It does not block any of the other changes request here. It would be fine to have a function that creates the default logger for now as long as other loggers can be added dynamically. The next step then will be to define multiple loggers in the configuration as outlined in the ticket. Those loggers can just be obtained by their name and will be created by the LogService class if necessary. Configured will be all loggers that we know about at this point. However the system should not be closed to additional loggers added from the outside.

**Comment by j3nsch** [comment]
Now the PR has shown up in Jira. Maybe because I fixed the name of the PR. I am not sure.

**Comment by j3nsch** [comment]
Please add a title to PRs in addition to OPUSVIER-nnnn. I know a lot of ticket numbers but it nevertheless helps me and other people that might have an interest in our development efforts. Thank you!

File: src/Log/LogService.php [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+            if (! is_dir($path)) {
+                throw new Exception('Directory for logging does not exist');
+            } else {
+                throw new Exception('Failed to open logging file:' . $logfilePath);
+            }
+        }
+
+    }
+
+    public function getConfig($configName){
```
Comment (j3nsch) [inline src/Log/LogService.php:75]: This is not going to work. There is only "Zend_Log" in the Zend_Registry, so getConfig() should be enough. However from that global configuration you need to get the config for a specific logger, so the function could handle that. It depends if you need the global config or just a subsection of it.

File: src/Bootstrap/Base.php [hunk 7]
```diff
@@ -134,76 +140,24 @@ protected function _initLogging()
...
+             if (! array_key_exists('SERVER_PROTOCOL', $_SERVER)
+                 and ! array_key_exists('REQUEST_METHOD', $_SERVER)) {
+                 $logFilename = "opus-console.log";
+             }
+         }
+
+        $logService = LogService::getInstance();
+        //        $logger = $logService->setDefaultLog($logFilename);
+
+        $logService->openLog('Zend_Config', $logFilename);
```
Comment (j3nsch) [inline src/Bootstrap/Base.php:156]: I think this should be more like
$logger = $logService->createLog('default', $logFilename);
With the log filename optional, but necessary here, because we want to (maybe) change it from the default "opus.log".

The next line with setLog() is hard to understand and should not be necessary. It behaves like getLog while the name is setLog?

File: test/LogTest.php [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        $this->logger = $thisLogInstance->setLog();    //how to get comparable log        
+    }
+
+    public function testSetLog(){
+        $log = $this->logInstance->setLog();
+        $this->assertSame($log, $this->logger);
+    }
+
+    public function testGetLog(){       
+        $log = $this->logInstance->getLog('opus.log');
```
[the remark this answers is not in the document]
Comment (j3nsch) [reply test/LogTest.php:58]: Please add more tests, for instance __testGetInstance__, even if they are empty for now. Please mark empty tests with $this->markTestIncomplete(), so they show up on test runs and won't be forgotten.

File: src/Bootstrap/Base.php (hunk 7 above)
Comment (j3nsch) [reply src/Bootstrap/Base.php:156]: Maybe a better name would be addLog. Hmm, not sure yet. I guess it depends, maybe we will have both.

File: src/Log/LogService.php [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        $formatter = new \Zend_Log_Formatter_Simple($this->format);
+
+        $writer = new \Zend_Log_Writer_Stream($this->logFile);	
+        $writer->setFormatter($formatter);
+
+        $logger = new \Zend_Log($writer);
+        return $logger;
+	}
+
+	public function setFormat($format = null){
```
Comment (j3nsch) [inline src/Log/LogService.php:173]: This function should be named __setDefaultFormat__, shouldn't it?

We also need a corresponding __getDefaultFormat__ function that is used even inside this class.

I think there are two separate things, the format string and creating the formatter object. I think creating the object should be placed in a function. I think we can do a bit more clean up here.

File: src/Log/LogService.php [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+ */
+
+namespace Opus\Log;
+
+class logService
+{	
+
+	private $logFile;
+	public static $logger = array(); 
+    public $format;
```
Comment (j3nsch) [inline src/Log/LogService.php:41]: Please write comments for all the variable, so it is clear what should be stored in them, strings, numbers or objects. Especially the objects are important. Please use "@var", because then the idea can pick it up and warn if there are assignments with the wrong type in the code. Thank you!

File: src/Log/LogService.php (hunk 6 above)
Comment (j3nsch) [reply src/Log/LogService.php:75]: Sorry, my mistake. It is still not going to work, but what you want is "Zend_Config". It is the global configuration for OPUS 4 read from `application.ini`, `config.ini` and `console.ini`. During tests there are also the `tests/config.ini` and the `tests/tests.ini`. The configuration for loggers would be a smaller part of the entire configuration.

File: src/Log/LogService.php [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    {
+        $this->config = \Zend_Registry::get('Zend_Config');
+    }
+
+    public function getPath()
+    {
+        return $this->logFilePath;
+    }
+
+    public function setPath($logFileName, $logPath = null)
```
Comment (j3nsch) [inline src/Log/LogService.php:96]: It would be good to have /** ... */ comments for these functions with a description of the intend. I don't think we need this function the way it is, but maybe I don't see the idea.

File: src/Log/LogServiceTest.php [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    public function testGetLog()
+    {
+        $log = $this->logInstance->getLog('opus.log');      //should pass only log name, not filename
+        $this->assertSame($log, $this->logger);
+    }
+
+    public function testSetDefaultLog()
+    {
+        $log = $this->logInstance->setDefaultLog('opus.log');
+        $this->assertSame($log, $this->logger);
```
Comment (j3nsch) [inline src/Log/LogServiceTest.php:73]: I am not sure what information we can extract from an `Zend_Log` object to verify that it is configured as expected. However we could manipulate the path for writing log files with a `setPath` function, write a defined message to the log file and verify it. It might make sense to write some generic functions for that because it might be useful for more than one test.

File: src/Log/LogService.php [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+                $this->openLog('Zend_Config', $logName);
+                return $this->addLog();
+            } catch {
+                throw new Exception("Cannot create new log", 1);
+
+            }
+        }
+    }
+
+    public function checkLogLevel()
```
[the remark this answers is not in the document]
Comment (j3nsch) [reply src/Log/LogService.php:174]: Actually looking at `addLog` basically almost all of the code from 127-153 with the exception of 147 could go here. The function should handle all that log level/priority stuff and free `addLog` from that responsibility. The code could be divided into two aspects __public__ `set/getDefaultLogLevel` and __private/protected__ `get/createLogLevelFilter`. One part handles the log level strings, the other creates the filter object. That we clean, however the second function might be overkill. We will see.

File: src/Log/LogService.php [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+    {
+        return isset($this->config->logging->log->level) ? $this->config->logging->log->level : false;
+    }
+
+    protected function createLog()
```
Comment (j3nsch) [inline src/Log/LogService.php:179]: Probably needs some parameters.

File: src/Log/LogService.php [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+        $writer = new \Zend_Log_Writer_Stream($this->logFile);
+        $writer->setFormatter($formatter);
+
+        $logger = new \Zend_Log($writer);
+        return $logger;
+    }
+
+    public function setDefaultFormat($format = null)		//DONE rename to setDefaultFormat()
+    {
+        $GLOBALS['id_string'] = uniqid(); // Write ID string to global variables, so we can identify/match individual runs.
```
Comment (j3nsch) [inline src/Log/LogService.php:196]: Would it make sense to move this code to `getDefaultFormat`, so you don't have to `setDefaultFormat` in order to get the __default__ format? Basically return the default format if nothing else has been set? That works similar to the getInstance function, but without any static stuff.

File: src/Log/LogService.php [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+            $invalidLogLevel = true;
+        }
+
+        // filter log output
+        $priorityFilter = new \Zend_Log_Filter_Priority($logLevel);
+        \Zend_Registry::set('LOG_LEVEL', $logLevel);
+        $logger->addFilter($priorityFilter);
+
+        if ($invalidLogLevel) {
+            $logger->err('Invalid log level \'' . $logLevelName .
```
Comment (j3nsch) [inline src/Log/LogService.php:151]: I assume this is from the old code. However since we want to improve it, try using "Invalid log level '$logLevelName' ..." (and so on). We normally use single quotes, they are safer. However when constructing strings like this double quotes and using the variable names inside the string make the code more readable.

File: src/Log/LogService.php [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+ * Foundation; either version 2 of the Licence, or any later version.
+ * OPUS is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details. You should have received a copy of the GNU General Public License
+ * along with OPUS; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ * @category    Framework
+ * @package     Opus_Bootstrap
```
Comment (j3nsch) [inline src/Log/LogService.php:28]: @category opus4-common
@package Opus\Log

There are still lots of incorrect annotations in the header. We want to be able to generate a nice API documentation. I am not sure about @category, but the name of the library seems like a good choice and could be replaced later.

File: src/Log/LogService.php [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+ * @author      Kaustabh Barman <barman@zib.de>
+ * @copyright   Copyright (c) 2008-2020, OPUS 4 development team
```
Comment (j3nsch) [inline src/Log/LogService.php:30]: File was created in 2020, so I think without "2008-". Thank you.

File: src/Log/LogServiceTest.php [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+        $this->logInstance = LogService::getInstance();
+        $this->logInstance->setPath('opus.log');        //may also mention new path as a second parametre
+        $this->logInstance->openLog('Zend_Config');
+
+        $this->logger = $thisLogInstance->setLog();    //how to get comparable log
+    }
+
+    public function testGetInstance()
+    {
+        $logInstance = new LogService();
```
Comment (j3nsch) [inline src/Log/LogServiceTest.php:54]: LogService should never be used this way. It is meant to be a singleton, so it should not be possible to use __new__. You can suppress that by having a private/protected __construct function. We haven't used that in PHP very much yet, so I do not know if there are any surprises waiting, but generally in OOP it should work that way.

Comment (j3nsch) [reply src/Log/LogServiceTest.php:54]: And you would expect a `testGetInstance` function to test `getInstance`.

Comment (kaustabhbarman) [reply src/Log/LogServiceTest.php:54]: Yes, I know it shouldn't be this way. Again, this commit wasn't meant to be reviewed, I left it like this just for me as a question, like, what if I do this, will correct it before a final commit for review. Sorry again for the confusion.

File: src/Log/LogService.php (hunk 15 above)
Comment (kaustabhbarman) [reply src/Log/LogService.php:196]: Yes, it makes sense. We can keep a `setFormat()` function for custom format. Then we can check if there's a custom format. If not, we can call `getDefaultFormat()`.

File: src/Log/LogService.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (kaustabhbarman) [reply src/Log/LogService.php:174]: Yes, it makes sense to make seperate functions. I think **public** `getDefaultLogLevel` and **protected** `createLogLevelFilter` are the suitable function names here. Though the later will just have the filter object and `addFilter` property of the logger passed in parameter called. We can always merge it later in the same function `setDefaultLogLevel` if it feels unnecessary.

File: src/Log/LogService.php [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+            }
+        }
+    }
+
+    public function getConfig()		//DONE not going to work
+    {
+        $this->config = \Zend_Registry::get('Zend_Config');
+    }
+
+    public function getPath()
```
[the remark this answers is not in the document]
Comment (kaustabhbarman) [reply src/Log/LogService.php:91]: I am keeping the `setPath()` function for the log folder path instead of file in my next commit. It will be managed in this class only.

File: src/Log/LogService.php [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+ * @package     Opus_Log
+ *
+ */
+class LogService
+{
+
+//DONE add comments for all variable -- use @var
+	/** @var Zend_Log $loggers */ 
+    public $loggers = [];		//all log objects with lognames as keys	
+    private $logFile;		//file pointer resource to the logfile
```
[the remark this answers is not in the document]
Comment (kaustabhbarman) [reply src/Log/LogService.php:48]: I can `fclose()` the pointer in `createlog()` function.

File: src/Log/LogService.php (hunk 11 above)
Comment (kaustabhbarman) [reply src/Log/LogService.php:96]: This function is to change the path of the logfile. It will be useful for running tests. I think you mentioned this on my previous commit.

File: src/Log/LogService.php [hunk 22]
```diff
@@ -0,0 +1,299 @@
...
+     * @param $logName string name of the log
+     * @param null $logFileName string optional name of the log file
+     */
+    public function createLog($logName, $logFileName = null)
+    {
+        $this->getConfig();
+
+        $logfilePath = $this->logPath == null ? $this->setPath($this->config->workspacePath . '/log/') : $this->logPath;
+
+        if ($logFileName == null) {
```
Comment (j3nsch) [inline src/Log/LogService.php:91]: This is tricky. The purpose of this function is not entirely clear yet to me. I think we need it and the way it is used right ow is basically okay. I think this bit about the logFileName needs work, but I think it will have wait until the rest has been addressed, because it gets to confusing otherwise. 

Basically one part has not been addressed at all yet, configured loggers. We have some global logging configuration, basically the defaults. We could more the default format into the configuration. That might be a nice improvement, but it isn't essential right now. However we want to add configuration for specific loggers later, so we have to ways of creating loggers. We can create them dynamically at runtime using this function. Do we need additional optional parameters for that, perhaps `format` and `priority/level`? The other way is to configure the translation logger with filename, format and level in the configuration. It a logger has been configured in this way it should be enough to call `getLog('translation')` and `LogService` will create the `Zend_Log` object using the specific and default configuration. We could decide that if *translation* has not been configured, we use all the default parameters except the file name. Using the file name `translation.log` makes sense as a backup if nothing else has been provided.

For now we could add a function `getLogConfig($name)`. It it returns null, we assume there is not specific configuration and we continue using the defaults. Another option here would be to basically always return something from `getLogConfig` and merge the default and specific configuration for a logger inside that function. That might make it very clean. However I do not think the rest of the code is there yet. So I suggest adding this function with a description simply returning *null* right now. We will add the configuration feature a little later when we have figured out the rest. It gets too confusing otherwise. Too many moving parts.

File: src/Bootstrap/Base.php [hunk 23]
```diff
@@ -139,70 +145,16 @@ protected function _initLogging()
...
-        }
-
-        if ($invalidLogLevel) {
-            $logger->err('Invalid log level \'' . $logLevelName .
-                    '\' configured.');
-        }
+        $logService = LogService::getInstance();
 
+        $logService->createLog('default', $logFilename);
+        $logger = $logService->addLog();
```
[the remark this answers is not in the document]
Comment (kaustabhbarman) [reply src/Bootstrap/Base.php:156]: I kept the `addLog()` function seperate from `createLog()` so that external modules can pass custom logs to `addLog()`. It is something that you mentioned in the comments of my previous commit.

[the remark this answers is not in the document]
Comment (j3nsch) [reply src/Bootstrap/Base.php:156]: The function createLog and addLog should both exit. However they shouldn't work the way they do right now and addLog should not be used here.

[the remark this answers is not in the document]
Comment (j3nsch) [reply src/Bootstrap/Base.php:156]: Oh, and I make mistakes, so if you see problems with my suggestions feel free to argue. Nevertheless `addLog` does not behave like I would expect it to and I think this will become more clear once the API of `LogService` is documented in the function and class comments.

File: src/Log/LogService.php (hunk 22 above)
Comment (kaustabhbarman) [reply src/Log/LogService.php:94]: We can create loggers during runtime without needing additional parameters like `format` or `priority/level` . But we can still configure the default parameters by calling `setDefaultFormat` and `setDefaultLogLevel()` (coming up on next commit). Also, calling `getlog('translation') will create a Zend_Log object with the default configurations. The file name will also be `translation.log` by default.

File: src/Log/LogService.php [hunk 24]
```diff
@@ -0,0 +1,368 @@
...
+    	return null;		
+    }
+
+    /**
+     *
+     * @param $logger \Zend_Log object
+     * @throws \Zend_Exception
+     * @throws \Zend_Log_Exception
+     *
+     * Gets log level constants from Zend_Log and apply log levels.
```
Comment (j3nsch) [inline src/Log/LogService.php:218]: This function does multiple things. A description with an "and" is a warning. But besides that, the function determines the default log level, which by the way is the log level configured unless there is nothing configured or it was set for the LogService manually. So all this should go into getDefaultLogLevel(). The function also handles verifying the log level string. It then has a  filter object created and adds to a logger. 

Currently it also sets the LOG_LEVEL in the global registry and this will happen more than once, perhaps with different levels, which wasn't the original intent. So as it is, this will create a bug, because there is other code relying on LOG_LEVEL being set to the level of the default logger.

**Comment by j3nsch** [comment]
Continued in PR #14
